packages/replay-next/components/elements-new/utils/serialization.ts

```diff
@@ -0,0 +1,268 @@
+  let stringTableKeyCount = 0;
+
+  const stringTableArray: number[] = [];
+  const stringTableKeysMap = new Map();
+
+  function getObjectId(value: any) {
+    if (typeof __RECORD_REPLAY_ARGUMENTS__ !== "undefined" && __RECORD_REPLAY_ARGUMENTS__ != null) {
+      // TODO [FE-2005][FE-2067] With persistent DOM ids, presumably we should switch to using this API?
+      // if (__RECORD_REPLAY_ARGUMENTS__.getPersistentId) {
+      //   const id = __RECORD_REPLAY_ARGUMENTS__.getPersistentId(value);
+      //   if (id) {
+      //     return id;
+      //   }
+      // }
+
+      if (
+        __RECORD_REPLAY_ARGUMENTS__.internal &&
+        __RECORD_REPLAY_ARGUMENTS__.internal.registerPlainObject
+      ) {
+        const id = __RECORD_REPLAY_ARGUMENTS__.internal.registerPlainObject(value);
+        if (id) {
+          return parseInt(id);
+        }
+      }
+    }
+
+    throw Error("Could not find object id");
+  }
+
+  function registerString(text: string | undefined) {
+    if (!text) {
+      return 0;
+    }
+    if (stringTableKeysMap.has(text)) {
+      return stringTableKeysMap.get(text);
+    } else {
+      const key = ++stringTableKeyCount;
+      stringTableKeysMap.set(text, key);
+
+      stringTableArray.push(key);
+      stringTableArray.push(text.length);
+
+      for (let charIndex = 0; charIndex < text.length; charIndex++) {
+        stringTableArray.push(text.charCodeAt(charIndex));
+      }
+
+      return key;
+    }
+  }
+
+  let nodesArray = [];
+  let queue: (number | Document | Element)[] = [0, rootNode];
+
+  while (queue.length > 0) {
+    const parentObjectId = queue.shift() as number;
+    const domNodeOrText = queue.shift() as Element;
+
+    const objectId = getObjectId(domNodeOrText);
+
+    let { children, classList, id, nodeType, tagName } = domNodeOrText;
+
+    switch (nodeType) {
+      case Node.DOCUMENT_NODE: {
+        tagName = "#document";
+        break;
+      }
+      case Node.DOCUMENT_TYPE_NODE: {
+        tagName = "<!Doctype>";
+        break;
+      }
+    }
+
+    let attributes: number[] = [];
+
+    // Attributes are serialized using the following format:
+    //   [ 0: stringKeyCount, 1: nameStringKey, ...valueStringKey ]
+    //
+    // For example, a name-only attribute (e.g. <div data-selected />) might be:
+    //   [ 1, 50 ]
+    //
+    // An id attribute (e.g. <div id="foo" />) might be:
+    //   [ 2, 63, 85 ]
+    //
+    // And a class name attribute (e.g. <div class="foo bar baz" />) might be:
+    //   [ 4, 22, 41, 42, 43 ]
+
+    // All node types should pass along id attribute
+    if (id) {
+      attributes.push(2);
+      attributes.push(registerString("id"));
+      attributes.push(registerString(id));
+    }
+
+    // All node types should pass along class attribute
+    // Class names are special; because of style reuse, we serialize individual class names separately
+    if (classList != null && classList.length > 0) {
```

Comment:
:speech_balloon: yeah, I figured it would be worth reusing classnames

Reply:
Given that #9943 is now sending all attributes _and_ text content, this ends up being an overkill optimization.